### ISSUE TITLE
Plottr crashing with keyerror

### ISSUE THREAD

**Issue by lakhotiaharshit** [toolsforexperiments/plottr#83]
I used the following code:
```python
import numpy as np
import qcodes as qc
from qcodes import (
    Measurement,
    experiments,
    initialise_or_create_database_at,
    new_experiment,
)
from qcodes.dataset.plotting import plot_dataset

station = qc.Station()

initialise_or_create_database_at("~/plottr_failing.db")

exp = new_experiment('dataset_context_manager2',
                                sample_name="no sample1")

meas = Measurement()
meas.write_period = 2 
N = 5

meas.register_custom_parameter(name='bfield',
                               label='B', unit='Tesla',
                               paramtype='numeric')
meas.register_custom_parameter(name='phase',
                               label='Phase', unit='rad',
                               paramtype='numeric',
                               setpoints=['bfield'])
meas.register_custom_parameter(name='time',
                               label='Time', unit='s',
                               paramtype='numeric')
meas.register_custom_parameter(name='signal',
                               label='Signal', unit='V',
                               paramtype='numeric',
                               setpoints=('time','bfield'))

with meas.run() as datasaver:
    for bf in np.linspace(0, 25, 10):
        bfield = bf
        phase = np.random.rand()
        time_data = np.linspace(0, 1, N)
        for time in time_data:
            signal_data = np.sin(2*np.pi*time) + 0.025*np.random.randn(N)
            datasaver.add_result(('time', time), ('signal', signal_data),('bfield',  bfield), ('phase', phase))
              
dataset = datasaver.dataset  # convenient to have for plotting
```
to generate dataset which looks like this: 
![image](https://user-images.githubusercontent.com/18750964/78293454-db809480-7528-11ea-84d2-60fc99b1df8a.png)

But Plottr crashes and gives the following error
```
Qcodes Logfile : C:\Users\a-halakh\.qcodes\logs\200402-41928-qcodes.log
2020-04-02 19:36:48,803 ¦ plottr ¦ ERROR ¦ grid ¦ process ¦ 299 ¦ Unknown data type <class 'plottr.data.datadict.DataDictBase'>.
2020-04-02 19:36:48,810 ¦ plottr ¦ ERROR ¦ grid ¦ process ¦ 299 ¦ Unknown data type <class 'plottr.data.datadict.DataDictBase'>.
2020-04-02 19:36:48,816 ¦ plottr ¦ ERROR ¦ grid ¦ process ¦ 299 ¦ Unknown data type <class 'plottr.data.datadict.DataDictBase'>.
Traceback (most recent call last):
  File "c:\users\a-halakh\documents\microsoft\plottr\plottr\apps\inspectr.py", line 438, in plotRun
    fc, win = autoplotQcodesDataset(pathAndId=(self.filepath, runId))
  File "c:\users\a-halakh\documents\microsoft\plottr\plottr\apps\autoplot.py", line 253, in autoplotQcodesDataset
    win = QCAutoPlotMainWindow(fc, pathAndId=pathAndId)
  File "c:\users\a-halakh\documents\microsoft\plottr\plottr\apps\autoplot.py", line 226, in __init__
    self.setDefaults()
  File "c:\users\a-halakh\documents\microsoft\plottr\plottr\apps\autoplot.py", line 189, in setDefaults
    self.fc.nodes()['Dimension assignment'].dimensionRoles = drs
  File "c:\users\a-halakh\documents\microsoft\plottr\plottr\node\node.py", line 40, in wrap
    self.ui.optSetters[optName](val)
  File "c:\users\a-halakh\documents\microsoft\plottr\plottr\node\dim_reducer.py", line 627, in setRoles
    self.widget.setRole(dimName, role)
  File "c:\users\a-halakh\documents\microsoft\plottr\plottr\node\dim_reducer.py", line 336, in setRole
    super().setRole(dim, role, **kw)
  File "c:\users\a-halakh\documents\microsoft\plottr\plottr\node\dim_reducer.py", line 276, in setRole
    super().setRole(dim, role)
  File "c:\users\a-halakh\documents\microsoft\plottr\plottr\node\dim_reducer.py", line 202, in setRole
    self.choices[dim]['roleSelectionWidget'].setCurrentText(role)
KeyError: 'bfield'
```

@astafan8

**Comment by wpfff** [comment]
Looking at the resulting table, i can see why reading this data might fail. In short, I think DataDict will interpret this as multiple data entries with the same coordinates (bfield and time both don't change over five entries). The error already happens in DataDict (expected!), and is then not properly caught and on the way the UI crashes (that's an issue). 

to fix this, an inner coordinate that has different values for each of the 5 entries for 'signal' would need to be added (could trivially be named 'repetition' or something like that).

The way DataDict is currently implemented the data here is faulty if i recall correctly. It might be an option to add a feature that automatically adds a missing axis, but so far i decided against that (because I wasn't certain about whether this is generally a good idea) and simply leave 'corrupt' data as unplottable.

**Comment by wpfff** [comment]
I should add: I phrased it incorrectly above. It's not the loading that fails, it's the gridding (can't work, for the reason i mentioned). For now, I'll look into making sure that the gridder won't do anything when it's presented with such data.

**Comment by jjwesd** [comment]
@wpfff to add to this as a feature request: if you plot this in plot_dataset from qcodes then it seems to work just fine. It will make a one-D plot for the phase and a 2D plot for the signal. In general I think it would be nice to have this functionality in plottr as well (e.g get dependend variables based on the data one selects in the data-selection window and determine the plot shape), since more often I want to have different dependencies for different variables 

![image](https://user-images.githubusercontent.com/8937555/79639819-4a4b2800-818e-11ea-92d2-67c6bbebc7ce.png)

An example of use cases i used it on already:
- save the readout frequency for every field value and then save the measured S parameter for every field value ánd driving frequency in two tone spectroscopy.
- Do vna traces of the resonator frequency(2D plots) versus applied current and meanwhile store the temperature that depends only on the current to see if you start heating up the MC

**Comment by wpfff** [comment]
@GitJaap what you're describing should work, though. does this not work in plottr?

In Harshit's case plottr is currently misbehaving (i think) because there are multiple data points with the same set of coordinates. As long as coordinates are unique (like in your case) that should work.
Let me know if not, then i should fix that.

**Comment by wpfff** [comment]
Note to self (or anyone else interested ;)):
The error reported by Harshit above (and similar ones) occurs when the data type switches (say, from grid data to other data) and widget elements in the dimension assignment widget aren't suited for the new data type. 
clearly a bug in that element, working on it.

**Comment by wpfff** [comment]
@lakhotiaharshit - i fixed the issue with that data (#96 ). Turns out the functionality that is needed for plotting this kind of data was already implemented, but there was a small bug due to a missing data type conversion.

@GitJaap - please still let me know if what plottr is doing now fulfills the need you described. to me it  sounds like it (at least after this fix :P), but i want to make sure.
